Add the first iteration of the Zinc API 1.0

jvican opened the pull request:
This is just the first iteration of the Zinc API.
  
This iteration's goal is to make safe functionality of Zinc public to allow Scala and Java parties (Gradle, Pants, Maven, et cetera) use Zinc with binary compatibility guarantees. This task needs another iteration given:
  
1. The big surface of the API; and,
2. The fact that the whole Zinc API was made private for 1.0 (!).

The last iteration will happen before RC1 (end of June - beginning of July).
  
The new public API makes it easier to:
  
* Get a compiled bridge for a given scala instance (via ivy).
* Get a scala instance (via ivy).
* Instantiate a Zinc compiler.
* Provide utilities for both Java and Scala users to get default implementations that do not expose the internals, placed in `sbt.internal.inc`.
  
Aside from these, this API also addresses some issues that have arisen in the past:
  
* It makes it easier to customize the classfile manager, as required by Scalajs and Scala Native. https://github.com/sbt/zinc/issues/125
* It replaces any instance of Maybe with Optional. https://github.com/sbt/zinc/issues/302
 
I find that in general, these commits make Zinc easier to use and improve the consistency of the whole API. The most remarkable change is the `CompileAnalysis` API, which now provides useful information for build tools like file stamps, compilations and source infos.
  
Note that all the utilities made public only expose functionality that is a product of the incremental compiler. No internal details of the incremental analysis are exposed to the users.
  
There are some Java files that have been moved out of the compiler interface because they required static method implementations in the projects on which the implementations of certain classes were provided.
  
In the future, I feel that these are the things that should be worked on to finish off the Zinc API:
  
1. Reorder project structure. Simplify it and make it more granular and easier to test. https://github.com/sbt/zinc/issues/286
2. Decide what and how to use to persist analysis files with appropriate benchmarking. https://github.com/sbt/zinc/issues/268 (protobuf is not possible, so we need to find a way to either use sbinary or a similar solution that does not harm (or even improve!) speed of compiler iterations).
3. Create utilities to instantiate:
  3.1. Empty `Analysis`.
  3.2. `DefinesClass`.
  3.3. `AnalysisStore` (up to discussion for next iteration).
  3.4. `JavaTool` et al, necessary to create `Compilers`.
  3.5. `PerClasspathLookupEntry`.
4. Make analysis content machine-independent: #218
  
I have more ideas, but those will be discussed when time's up.

The breaking changes are documented in every commit. I suggest that reviewers go through the commits one by one, though I don't expect them to do so (given the big diff size). As this is just the first iteration of the redesign and targets a milestone, I am not writing any migration notes for now. I'll wait until everything is finished in our first RC to write the migration notes so that other build tools (other than sbt) can start using it.
  
As a reminder, the public API is `xsbti`. Scala implementations live in `sbt.internal.inc`.
  
This is a good time to gather feedback. So that's very much welcome.

jvican committed:
Remove deprecated compatibility stubs

Remove unused `GlobalCompat` for compatibility with 2.8.1 and extract
the run subclassing out of the main logic of the `run` method.

jvican committed:
Make Zinc `Global` implementations independent

The callback implementation should be independent of the compiler
interface and the source code should reflect so.

Breaking change in comparison with the previous API: `Compiler` is now
called `ZincCompiler` to make the naming more clear and better
distinguish between `CachedCompiler` and its underlying `ZincCompiler`.

Docs have been added and simplifications to methods have been provided
and double checked with previous Scala versions. For instance, `dropRun`
has been confirmed to only exist in 2.10.x series and removed in 2.11.x.
Therefore, its implementation has been moved to the compatibility traits
for 2.10.x series.  We then remove a reflective call for versions above
2.10.6.

jvican committed:
Hide `compilation` from `AnalysisCallback`

jvican committed:
Port `CompileAnalysis` from contraband to Java

We need `CompileAnalysis` to start creating the Java API with good docs.
Contraband is not necessary for the analysis since we only
serialize/deserialize `MAnalysis`, its subclass.

jvican committed:
Change `Stamps` interface and functionality

* Make `ReadStamps` a Java interface exposed in the public API.
* Create a `Stamp` interface that Java users can use without any
  knowledge of the internal requirement.
* Add docs to the pertinent Java interface.
* Change `Stamp` companion object by `Stamper` to make it Java-friendly.
  I also think the name is clearer about the contents of the object.
* Rename `internalSouce` to `source` for consistency with the Scala API.
  It's made clear that source should always be an internal source in the
  docs. This makes sense anyway, otherwise it wouldn't make sense to ask
  for the stamp of a source file that is not compiled in the current
  analysis.

jvican committed:
Remove unused stamp `Exists`

`Exists` has never been used, neither internally in Zinc nor in sbt and
pants (I double checked on both).

The concept of `exists` does not make sense, since there is no value in
asserting that a stamp "exists" only in asserting that a stamp is
**not** available for a given file.

The latter is represented by the new `Stamp` API since last modified
time and hash are both optional. If both are empty, a stamp does not
exist.

jvican committed:
Turn `CompileAnalysis` to an interface

`CompileAnalysis` should be an interface instead of an abstract class.

jvican committed:
Add `ReadStamps` to `CompileAnalysis`

Users of `CompileAnalysis` will have access to the stamps of the files
that have been processed by Zinc.

jvican committed:
Create a `SourceInfo` API to be consumed from Java

jvican committed:
Create `ReadSourceInfos` & add to `CompileAnalysis`

`ReadSourceInfos` is a read-only interface for getting source info
(namely, compiler errors) associated with a given file.

This is very useful in lots of cases. Especially to process and improve
error messages in a lightweight way, though ScalaClippy already exists
for that purpose.

In any case, having access to the compiler errors outside of is of
ultimate importance for tools that want to preprocess compilation
results independently of sbt.

jvican committed:
Clarify `Stamp` documentation for empty value

jvican committed:
Replace `OutputSetting` by `OutputGroup`

`OutputSetting` is like `OutputGroup` by just returns strings instead of
files. I see no reason why `OutputGroup` should not be used instead,
therefore simplifying the Zinc API and allowing users to have access to
files and not mere strings. If they want strings, they always can do
`getAbsolutePath`.

N.B. This is good for machine independence.

jvican committed:
Fix `Output` and `Compilation` API

* Remove dangerous use of output groups as fields in compilation. In the
  case of `SingleOutput`, the source directory was invented and
  incorrect (the root). Users of the API processing this file could
  create chaos in people's computers. The new design forces consumers
  of the `Compilation` API to make up their minds and identify whether
  the compilation was used with a single output or several outputs.
  In order to do that, we return `Output` instead of an array of
  `OutputGroup`.
* Rename `outputDirectory` and `sourceDirectory` to be Java friendly.
* Augment `Output` interface with Java-friendly methods. Scala users
  can use pattern matching instead.

jvican committed:
Remove `apiNote` since Javadoc doesn't accept it

jvican committed:
Make the Output API more Java friendly

* Make methods in `Output` subclasses public
* Changes to the API. Java idiomatic renames:
  * sourceDirectory -> getSourceDirectory
  * outputDirectory -> getOutputDirectory

jvican committed:
Add `ReadCompilations` to `CompileAnalysis`

* `Compilation` has been ported to Java code and documented.
* `ReadCompilations` is a read-only API of `Compilations`.
* Add a TODO to check the ordering of the array of compilations.
* Expose `ReadCompilations` to the `CompileAnalysis`.

jvican committed:
Add `CompilerBridgeProvider` interface and misc.

* Add interface for the provider.
* Rename `IvyComponentCompiler` to `ZincComponentCompiler`.
* Split `ZincComponentCompiler` and `ZincComponentManager`.
* Define `IfMissing` and `InvalidComponent` into independent files.
* Rename variables and internal API to be clearer.

jvican committed:
Move `constant` to static in `CompilerBridgeProvider`

No need to have it in the internal Scala side.

jvican committed:
Add headers to files that don't have them

jvican committed:
Rename `getCompilerSources` to `getCompiledBridge`

Previous name was incorrect, we actually return the compiled bridge.

jvican committed:
Fix oversight in InvalidComponent message

jvican committed:
Add `getScalaInstance` to `CompilerBridgeProvider`

jvican committed:
Remove custom `getScalaInstance` and use default

jvican committed:
Test `getScalaInstance` & `getCompiledBridge`

* Remove unused code in `BridgeProviderSpecification` and test the real
  code in the default provider.
* Don't use temporary directory to download the Scala jars, retrive them
  in a directory specified by the user.
* Use `java.net.ClassLoader` instead of `ScalaClassLoader`.
* Use the `ScalaInstance` interface, not the implementation. Remove any
  reference to the implementation.

jvican committed:
Compile bridge for 2.12.x and other Scala versions

This commit makes sure that we always check that we compile against the
latest two versions in Scala 2.10, Scala 2.11 and Scala 2.12.

This commit also removes the null loggers and uses normal loggers to
check what the output of the bridge provider is. This output is
important and the rationale to do this is that it should be visible for
us in order to detect cases where there's a regression or the output is
modified in some way. Output is short anyway.

jvican committed:
Fix error when showing unresolved warning lines

jvican committed:
Add Java API for Zinc Ivy integration

* Adds a method to fetch the default implementation of the compiler
  provider.
* Adds a method to get a default `IvyConfiguration` from a small number
  of parameters.

The public API is implemented in Java to ensure binary compatibility no
matter what happens to the internal Scala implementations. It also makes
it easier for Java users to use the API and avoid importing MODULE$'s
and Scala generated class files.'

jvican committed:
Add public method to get a local ivy resolver

jvican committed:
Add public method to get no-op global lock

jvican committed:
Remove verbose Java global lock implementation

jvican committed:
Add public method to get default component provider

Expose default component provider through a static method in
`ZincCompilerBridge`. This method is necessary to invoke the bridge.

Removes the previous global lock and component provider implementations
in `BridgeProviderSpecification`.

jvican committed:
Remove the use of `BaseIvySpecification`

This is the proof that we're now testing the Zinc public API because
we're resolving with the ivy options and resolution specified in the
default compiler bridge provider.

jvican committed:
Move IncOptionsUtil to zincCore project

It does not make sense to define it in the compiler interface since the
default implementations of class file manager and some inc options are
defined in `zincCore`.

jvican committed:
Format `IncOptionsUtil`

jvican committed:
Fix `fromStringMap` for classfile manager key

Previously, classfile manager key was not read from
`incOptions.properties` files because there was no way to specify which
classfile manager should be chosen.

Now, the class file manager can be indeed chosen. As we cannot serialize
the logger that should be used for the transactional classfile manager
type, we reuse the one passed in at the time of reading the incremental
options values.

jvican committed:
Add docs to `IncToolOptionsUtil`

jvican committed:
Fix #302: Migrate from Maybe to Optional

jvican committed:
Make the equiv for stamp public

This is necessary for other users to implement external hooks and
reason about which files have been changed and which have not.

Java users can access it with:

```
sbt.internal.inc.Stamp$.MODULE$.equivStamp
```

jvican committed:
Add interface for external Lookup

* Move `JavaInterfaceUtil` to zincCore.
* Define `Changes` in `compilerInterface`.
* Define a `Lookup` interface.
  Note that this lookup interface can now be implemented by third
  parties because the stamps of the files are provided by the
  `CompileAnalysis` interface.

jvican committed:
Replace null by Optional in ExternalHooks

For God's sake, no nulls.

jvican committed:
Fix #125: Get default classfile manager for inc

Previously, there was no way to get the underlying classfile manager
without depending on internal implementation details (sbt.internal.inc).

This problem is described in this issue: https://github.com/sbt/zinc/issues/125.
That is a blocker for scalajs.

* Add `ClassFileManagerUtil` static methods to get access to the
  underlying class file manager implementation from both the key and the
  incremental options.
* Add docs to subclasses of `ClassFileManagerType` and the
  implementations.
* Remove closures for getting the `deleteImmediately` and
  `transactional` class file manager implementations.
* Extract implementation from anonymous classes, make them final and
  private.

jvican committed:
Remove deprecated `transactional`

In sbt 0.13.x series, there exist a transactional classfile manager that
did not have any logger on which to log its steps.

As this is not used anymore, and a classfile manager should always log
them, we remove this from the internal API.

jvican committed:
Remove `allInternalSrcDeps` from Relations

It's safe to remove this deprecated method in M1 because it's part of
the private API (sbt.internal.inc).

This method has been deprecated and removed because class-based
dependency tracking does not need to know about internal source
dependencies anymore (in fact, this is not a useful concept for
class-based dependency tracking).

jvican committed:
Add `ZincCompilerUtil` to get the Scalac compiler

* Move `IncrementalCompilerUtil` to `ZincCompilerUtil`.
* Change `IncrementalCompilerUtil` to an interface with static methods.
* Remove bug in which ScalaInstance implementation was leaking to the
  public API.

jvican committed:
Hide Zinc Compilation implementation

Before, we had made this public in `sbt.inc`, but this is not necessary
anymore.

jvican committed:
Move formats away from public API

Formats for `AnalyzedClass`, `Compilation` and `Companions` were in the
public API (xsbti.api)! They have been removed from there.

jvican committed:
Move ZincBridgeProvider to `xsbti.compile`

Before, we were using `sbt.inc`. As we have settled to define all the
interfaces in the `xsbti.compile` namespaces, it's not worth it to force
users remember other locations for the interfaces forwarding the default
implementations in Scala.

jvican committed:
Move `ClasspathOptionsUtil` to the Java public API

Move the previous Scala implementation to Java as static methods.

jvican committed:
Add headers to files missing it

jvican committed:
Make `CompilerCache` public

eed3si9n commented:
As a test, could we put `zinc/src/test/scala/sbt/inc/*Spec.scala` into another package name like `com.example`, and see if it's able to initialize everything without invoking the internals?

While I agree with getting rid of `Maybe` and using `Optional` it's a bit distracting to review it because it spans across everywhere. How difficult do you think it would be to split Maybe change into a another PR that precedes this?

eed3si9n commented on internal/zinc-scripted/src/main/contraband-scala/sbt/internal/inctest/Build.scala:
```diff
@@ -1,32 +0,0 @@
-
-// DO NOT EDIT MANUALLY
-package sbt.internal.inctest
-final class Build private (
```
What happened to my build.json?

jvican replied:
This has been recovered.

eed3si9n commented on internal/compiler-interface/src/main/java/xsbti/compile/CompilerBridgeProvider.java:
```diff
@@ -0,0 +1,20 @@
+     * @param logger The logger.
+     * @return The jar (more typically) or directory where the sources are stored.
+     */
+    File getBridgeSources(ScalaInstance scalaInstance, Logger logger);
```
Getter shouldn't have side effects.

jvican replied:
I'll call it `fetchBridgeSources`.

eed3si9n commented on internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/IfMissing.scala:
```diff
@@ -0,0 +1,14 @@
+package sbt.internal.inc
+
+sealed trait IfMissing
```
Not sure why this needs to be on its own file.

jvican replied:
We may want to explore to make this stable in the next iteration. That's why I split it up, there was too much logic in its place of origin.

eed3si9n commented on internal/compiler-interface/src/main/java/xsbti/compile/CompilerBridgeProvider.java:
```diff
@@ -22,7 +22,7 @@
         return new CompilerBridgeProvider() {
             @Override
-            public File getBridgeSources(ScalaInstance scalaInstance, Logger logger) {
+            public File getCompiledBridge(ScalaInstance scalaInstance, Logger logger) {
```
Same as before. Getter shouldn't have side effects.

eed3si9n reviewed (changes requested):
Some minor comments.

jvican commented:
> As a test, could we put zinc/src/test/scala/sbt/inc/*Spec.scala into another package name like com.example, and see if it's able to initialize everything without invoking the internals?
  
This is a good idea. I'ld like to do this in the second iteration, since there are still some things that must be got from `internal`. Those I specify in the PR description, but I may be missing more.
  
> While I agree with getting rid of Maybe and using Optional it's a bit distracting to review it because it spans across everywhere. How difficult do you think it would be to split Maybe change into a another PR that precedes this?
  
The diff for the maybe migration is small, see https://github.com/sbt/zinc/pull/304/commits/686b6bb174fe76390646a7e7770914906fefc2f1. For time-consuming reasons, I would prefer to get it merged in this PR.

jvican committed:
Rename `getXXX` by `fetchXXX`

As suggested by Eugene.

jvican commented:
Feedback has been addressed @eed3si9n.

jvican commented:
Tests are green.

eed3si9n commented:
Thanks for all the work on this.

stuhood commented:
With regard to additional private/internal APIs that are currently in use, can browse through https://github.com/pantsbuild/pants/tree/master/src/scala/org/pantsbuild/zinc (as of zinc `1.0.0-X7`) ... I'm currently working on updating to include this patch.

jvican commented:
@stuhood I've been browsing through that for a while, but thanks for the heads up. Are you in a hurry to do this? I'll be tackling the second iteration of the API soon before the release candidate and the API freeze, so maybe you want to wait until that's out?

eed3si9n commented:
The change I made to sbt is https://github.com/sbt/sbt/pull/3228

stuhood commented:
@jvican : It would probably make sense for me to continue to attempt this, just in order to find any additional issues that you might not have seen. As an example, yesterday I noticed that it's not currently possible to extend `LoggerReporter` in a useful way in `X16`, because various methods were made private. So https://github.com/pantsbuild/pants/blob/master/src/scala/org/pantsbuild/zinc/logging/Reporters.scala#L28 is no longer possible. A theoretical `FilteredReporter` wrapper might make more sense anyway though.

jvican commented:
> So https://github.com/pantsbuild/pants/blob/master/src/scala/org/pantsbuild/zinc/logging/Reporters.scala#L28 is no longer possible. A theoretical FilteredReporter wrapper might make more sense anyway though.
  
Great, I'll make this (or something similar, I have to look into the details) happen to unblock that implementation.

stuhood commented:
@jvican: A few more details here:
1. the Reporter API seems to require a `ManagedLogger`, but it's not clear how (or whether it would be a good idea) to create one of those, versus passing in an `xsbti.Logger` or `sbt.util.Logger`
2. Because a `Reporter` is a required parameter for `IncrementalCompilerImpl.compile` (and there doesn't appear to be a way to create a `LoggerReporter` without a `ManagerLogger`), it does not currently seem to possible to compile using the 18 arg form.

I'm going to try porting to usage of `xsbti.compile.Inputs` instead.

stuhood commented:
Alright, I ported to `xsbti.compile.Inputs`, and it cleared things up considerably. But the above issue with `LoggerReporter` and `ManagedLogger` is still an issue.

Also, it doesn't look like it's possible to use a "default" Reporter (as it is with `Progress`, which is `Optional`)... it's conceivable that at some point in the future we'll move the log filtering out of the Reporter, at which point being able to use a default by specifying `None` would be good.

jvican commented:
@stuhood: Yes, it seems that the entanglement between `LoggerReporter` and `ManagedLogger` is due to sbt. It's an API where sbt's design decisions leak too much.
  
To overcome this issue, some of Zinc's test suites create a `ConsoleReporter` that is a direct subclass of `Reporter` and just `println`s, but that's far from desirable. I think I am going to create a `ManagedLogger` for Zinc and expose it somewhere, or at least make it trivial to create a Zinc reporter. I am working on it as we speak.
  
With regards to making the reporter `Optional`, I have my doubts this is the best choice. It's such a key component of the compilation that I would prefer that users explicitly pass it. Also, depending on the use case, it's likely to be changed. What is a safe assumption for the user? Will Zinc output to my console? Since it's not obvious, I believe users may be better off being more explicit about what they want. I guess that you need for this would disappear if we add methods to instantiate zinc reporters that you can pass in. WDYT?

jvican commented:
For the record, if you want to move forward and need a `ManagedLogger`, you can do `sbt.util.LoggerExchange.logger(...)`.

stuhood commented:
> What is a safe assumption for the user? Will Zinc output to my console? Since it's not obvious, I believe users may be better off being more explicit about what they want. I guess that you need for this would disappear if we add methods to instantiate zinc reporters that you can pass in. WDYT?

Yea, that makes sense. I'm still thinking of zinc as a CLI tool, but in the context of a library, being explicit about output location is important... ditto logging.

jvican commented:
Hey @stuhood, I've been fixing all the issues you've reported regarding logging. I think I have come up with some elegant solutions you may want to have a look at.

This is the most important change: https://github.com/scalacenter/zinc/commit/a9a9c2faf3b511c0f95b19fab87eea2e67346408.
  
All the changes are in this branch: https://github.com/sbt/zinc/compare/1.0...scalacenter:fix-loggers.
  
I can release these changes as part of a temporary `scalacenter/zinc` release to unblock you and get more feedback. These changes will need, of course, to be merged upstream in sbt/zinc.
  
In the new interface, I have added a filtered reporter in Zinc and the normal `LoggerReporter` (now renamed to `LoggedReporter`, see https://github.com/scalacenter/zinc/commit/cac014918ec6d4ff20b257d8363256739d5645e1). `LoggedReporter` now accepts `xsbti.Logger`, meaning that you can pass in whatever logging mechanism you want so long as it extends that interface. These reporter implementations are internal, but you can get a reporter backed up by a logger via the public `ReporterUtil`: https://github.com/scalacenter/zinc/commit/60820b18b7d56f0b7c9e9a9ae267c2e5bc50b59c.
  
Let me know if this public API works for you.

stuhood commented:
Thanks @jvican . The API looks great (with one exception mentioned below): thanks for incorporating file/msg based filtering.

I should be able to sanity check incremental compile before the end of Monday.

There is a bit of inconsistency between `java.util.function.Function` and `xsbti.F1` for positionMappers in `ReporterUtil` and `CompileOptions`. pants currently only requires JDK7, but I'm floating the idea of switching to JDK8 now... `Function` should be fine, but consistency would be good.

jvican commented:
@stuhood Happy that you mention that, I opened a PR in sbt/util to address that issue three days ago. The idea is to remember any trace of the old Java-friendly classes for bincompat (`F0`, `F1` and `Maybe`) and replace them by the default Java ones (`Supplier`, `Function` and `Optional`).
https://github.com/sbt/util/pull/84

stuhood commented:
The new API looks good from a logging/usability perspective (other than #317). I have things working well enough to have reproduced https://github.com/pantsbuild/pants/issues/4477, so I will begin debugging that tomorrow.

stuhood commented:
@jvican : If you have a PR for the `Reporters`/`Logger` change, I'd shipit. I'm getting close to needing an `X17` release to move forward.

jvican commented:
Shipping it tomorrow morning. It's ready but kept it local because I'm changing more things in the API.
